From 739e79cf4bde70a198bb5f79a465186ec136f491 Mon Sep 17 00:00:00 2001
From: Yanan Wang <wangyanan55@huawei.com>
Date: Tue, 6 Feb 2024 21:34:55 +0800
Subject: [PATCH 510/960] downstream: KVM: arm64: Only probe Hisi ncsnp feature
 on Hisi CPUs

The "ncsnp" is an implementation specific CPU virtualization
feature on Hisi 1620 series CPUs. This feature works just
like ARM standard S2FWB to reduce some cache management
operations in virtualization.
Given that it's Hisi specific feature, let's restrict the
detection only to Hisi CPUs. To realize this:
1) Add a sub-directory `hisilicon/` within arch/arm64/kvm to
hold code for Hisi specific virtualization features.
2) Add a new kconfig option `CONFIG_KVM_HISI_VIRT` for users
to select the whole Hisi specific virtualization features.
3) Add a generic global KVM variable `kvm_ncsnp_support`
which is `false` by default. Only re-initialize it when
we have `CONFIG_KVM_HISI_VIRT` enabled.

Signed-off-by: Yanan Wang <wangyanan55@huawei.com>
Signed-off-by: lishusen <lishusen2@huawei.com>
Signed-off-by: Slim6882 <15605196882@163.com>
Signed-off-by: chenyi <chenyi211@h-partners.com>
Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 arch/arm64/include/asm/hisi_cpu_model.h |  21 ----
 arch/arm64/include/asm/kvm_host.h       |   2 +-
 arch/arm64/kernel/image-vars.h          |   5 +
 arch/arm64/kvm/Kconfig                  |   1 +
 arch/arm64/kvm/Makefile                 |   2 +-
 arch/arm64/kvm/arm.c                    |  13 ++-
 arch/arm64/kvm/hisi_cpu_model.c         | 117 ----------------------
 arch/arm64/kvm/hisilicon/Kconfig        |   7 ++
 arch/arm64/kvm/hisilicon/Makefile       |   2 +
 arch/arm64/kvm/hisilicon/hisi_virt.c    | 124 ++++++++++++++++++++++++
 arch/arm64/kvm/hisilicon/hisi_virt.h    |  19 ++++
 config.aarch64                          |   1 +
 config.aarch64-64k                      |   1 +
 13 files changed, 171 insertions(+), 144 deletions(-)
 delete mode 100644 arch/arm64/include/asm/hisi_cpu_model.h
 delete mode 100644 arch/arm64/kvm/hisi_cpu_model.c
 create mode 100644 arch/arm64/kvm/hisilicon/Kconfig
 create mode 100644 arch/arm64/kvm/hisilicon/Makefile
 create mode 100644 arch/arm64/kvm/hisilicon/hisi_virt.c
 create mode 100644 arch/arm64/kvm/hisilicon/hisi_virt.h

diff --git a/arch/arm64/include/asm/hisi_cpu_model.h b/arch/arm64/include/asm/hisi_cpu_model.h
deleted file mode 100644
index 67008d17416e..000000000000
--- a/arch/arm64/include/asm/hisi_cpu_model.h
+++ /dev/null
@@ -1,21 +0,0 @@
-/* SPDX-License-Identifier: GPL-2.0-or-later */
-/*
- * Copyright(c) 2019 Huawei Technologies Co., Ltd
- */
-
-#ifndef __HISI_CPU_MODEL_H__
-#define __HISI_CPU_MODEL_H__
-
-enum hisi_cpu_type {
-	HI_1612,
-	HI_1616,
-	HI_1620,
-	UNKNOWN_HI_TYPE
-};
-
-extern enum hisi_cpu_type hi_cpu_type;
-extern bool kvm_ncsnp_support;
-
-void probe_hisi_cpu_type(void);
-void probe_hisi_ncsnp_support(void);
-#endif /* __HISI_CPU_MODEL_H__ */
diff --git a/arch/arm64/include/asm/kvm_host.h b/arch/arm64/include/asm/kvm_host.h
index 35bfd01c9160..6b5d17aa3dda 100644
--- a/arch/arm64/include/asm/kvm_host.h
+++ b/arch/arm64/include/asm/kvm_host.h
@@ -27,7 +27,6 @@
 #include <asm/fpsimd.h>
 #include <asm/kvm.h>
 #include <asm/kvm_asm.h>
-#include <asm/hisi_cpu_model.h>
 
 #define __KVM_HAVE_ARCH_INTC_INITIALIZED
 
@@ -1196,5 +1195,6 @@ void kvm_arm_vcpu_power_off(struct kvm_vcpu *vcpu);
 bool kvm_arm_vcpu_stopped(struct kvm_vcpu *vcpu);
 
 extern bool force_wfi_trap;
+extern bool kvm_ncsnp_support;
 
 #endif /* __ARM64_KVM_HOST_H__ */
diff --git a/arch/arm64/kernel/image-vars.h b/arch/arm64/kernel/image-vars.h
index 35f3c7959513..68380ea092ae 100644
--- a/arch/arm64/kernel/image-vars.h
+++ b/arch/arm64/kernel/image-vars.h
@@ -106,6 +106,11 @@ KVM_NVHE_ALIAS(__hyp_rodata_end);
 /* pKVM static key */
 KVM_NVHE_ALIAS(kvm_protected_mode_initialized);
 
+#ifdef CONFIG_KVM_HISI_VIRT
+/* Capability of non-cacheable snooping */
+KVM_NVHE_ALIAS(kvm_ncsnp_support);
+#endif
+
 #endif /* CONFIG_KVM */
 
 #ifdef CONFIG_EFI_ZBOOT
diff --git a/arch/arm64/kvm/Kconfig b/arch/arm64/kvm/Kconfig
index 83c1e09be42e..8350d43f56d4 100644
--- a/arch/arm64/kvm/Kconfig
+++ b/arch/arm64/kvm/Kconfig
@@ -5,6 +5,7 @@
 
 source "virt/lib/Kconfig"
 source "virt/kvm/Kconfig"
+source "arch/arm64/kvm/hisilicon/Kconfig"
 
 menuconfig VIRTUALIZATION
 	bool "Virtualization"
diff --git a/arch/arm64/kvm/Makefile b/arch/arm64/kvm/Makefile
index 86c6ded87eeb..826a05d072d7 100644
--- a/arch/arm64/kvm/Makefile
+++ b/arch/arm64/kvm/Makefile
@@ -15,7 +15,6 @@ kvm-y += arm.o mmu.o mmio.o psci.o hypercalls.o pvtime.o \
 	 guest.o debug.o reset.o sys_regs.o stacktrace.o \
 	 vgic-sys-reg-v3.o fpsimd.o pkvm.o \
 	 arch_timer.o trng.o vmid.o emulate-nested.o nested.o \
-	 hisi_cpu_model.o \
 	 vgic/vgic.o vgic/vgic-init.o \
 	 vgic/vgic-irqfd.o vgic/vgic-v2.o \
 	 vgic/vgic-v3.o vgic/vgic-v4.o \
@@ -24,6 +23,7 @@ kvm-y += arm.o mmu.o mmio.o psci.o hypercalls.o pvtime.o \
 	 vgic/vgic-its.o vgic/vgic-debug.o
 
 kvm-$(CONFIG_HW_PERF_EVENTS)  += pmu-emul.o pmu.o
+obj-$(CONFIG_KVM_HISI_VIRT) += hisilicon/
 
 always-y := hyp_constants.h hyp-constants.s
 
diff --git a/arch/arm64/kvm/arm.c b/arch/arm64/kvm/arm.c
index efe9cb108651..0b82f31b2142 100644
--- a/arch/arm64/kvm/arm.c
+++ b/arch/arm64/kvm/arm.c
@@ -47,6 +47,10 @@
 
 static enum kvm_mode kvm_mode = KVM_MODE_DEFAULT;
 
+#ifdef CONFIG_KVM_HISI_VIRT
+#include "hisilicon/hisi_virt.h"
+#endif
+
 DECLARE_KVM_HYP_PER_CPU(unsigned long, kvm_hyp_vector);
 
 DEFINE_PER_CPU(unsigned long, kvm_arm_hyp_stack_page);
@@ -56,8 +60,7 @@ DECLARE_KVM_NVHE_PER_CPU(struct kvm_cpu_context, kvm_hyp_ctxt);
 
 static bool vgic_present, kvm_arm_initialised;
 
-/* Hisi cpu type enum */
-enum hisi_cpu_type hi_cpu_type = UNKNOWN_HI_TYPE;
+/* Capability of non-cacheable snooping */
 bool kvm_ncsnp_support;
 
 static DEFINE_PER_CPU(unsigned char, kvm_hyp_initialized);
@@ -2460,9 +2463,11 @@ static __init int kvm_arm_init(void)
 		return err;
 	}
 
-	/* Probe the Hisi CPU type */
+#ifdef CONFIG_KVM_HISI_VIRT
 	probe_hisi_cpu_type();
-	probe_hisi_ncsnp_support();
+	kvm_ncsnp_support = hisi_ncsnp_supported();
+#endif
+	kvm_info("KVM ncsnp %s\n", kvm_ncsnp_support ? "enabled" : "disabled");
 
 	in_hyp_mode = is_kernel_in_hyp_mode();
 
diff --git a/arch/arm64/kvm/hisi_cpu_model.c b/arch/arm64/kvm/hisi_cpu_model.c
deleted file mode 100644
index 52eecf1ba1cf..000000000000
--- a/arch/arm64/kvm/hisi_cpu_model.c
+++ /dev/null
@@ -1,117 +0,0 @@
-// SPDX-License-Identifier: GPL-2.0-or-later
-/*
- * Copyright(c) 2019 Huawei Technologies Co., Ltd
- */
-
-#include <linux/acpi.h>
-#include <linux/of.h>
-#include <linux/init.h>
-#include <linux/kvm_host.h>
-
-#ifdef CONFIG_ACPI
-
-/* ACPI Hisi oem table id str */
-const char *oem_str[] = {
-	"HIP06",	/* Hisi 1612 */
-	"HIP07",	/* Hisi 1616 */
-	"HIP08"		/* Hisi 1620 */
-};
-
-/*
- * Get Hisi oem table id.
- */
-static void acpi_get_hw_cpu_type(void)
-{
-	struct acpi_table_header *table;
-	acpi_status status;
-	int i, str_size = ARRAY_SIZE(oem_str);
-
-	/* Get oem table id from ACPI table header */
-	status = acpi_get_table(ACPI_SIG_DSDT, 0, &table);
-	if (ACPI_FAILURE(status)) {
-		pr_err("Failed to get ACPI table: %s\n",
-		       acpi_format_exception(status));
-		return;
-	}
-
-	for (i = 0; i < str_size; ++i) {
-		if (!strncmp(oem_str[i], table->oem_table_id, 5)) {
-			hi_cpu_type = i;
-			return;
-		}
-	}
-}
-
-#else
-static void acpi_get_hw_cpu_type(void) {}
-#endif
-
-/* of Hisi cpu model str */
-const char *of_model_str[] = {
-	"Hi1612",
-	"Hi1616"
-};
-
-static void of_get_hw_cpu_type(void)
-{
-	const char *cpu_type;
-	int ret, i, str_size = ARRAY_SIZE(of_model_str);
-
-	ret = of_property_read_string(of_root, "model", &cpu_type);
-	if (ret < 0) {
-		pr_err("Failed to get Hisi cpu model by OF.\n");
-		return;
-	}
-
-	for (i = 0; i < str_size; ++i) {
-		if (strstr(cpu_type, of_model_str[i])) {
-			hi_cpu_type = i;
-			return;
-		}
-	}
-}
-
-void probe_hisi_cpu_type(void)
-{
-	if (!acpi_disabled)
-		acpi_get_hw_cpu_type();
-	else
-		of_get_hw_cpu_type();
-
-	if (hi_cpu_type == UNKNOWN_HI_TYPE)
-		pr_warn("UNKNOWN Hisi cpu type.\n");
-}
-
-#define NCSNP_MMIO_BASE	0x20107E238
-
-/*
- * We have the fantastic HHA ncsnp capability on Kunpeng 920,
- * with which hypervisor doesn't need to perform a lot of cache
- * maintenance like before (in case the guest has non-cacheable
- * Stage-1 mappings).
- */
-void probe_hisi_ncsnp_support(void)
-{
-	void __iomem *base;
-	unsigned int high;
-
-	kvm_ncsnp_support = false;
-
-	if (hi_cpu_type != HI_1620)
-		goto out;
-
-	base = ioremap(NCSNP_MMIO_BASE, 4);
-	if (!base) {
-		pr_err("Unable to map MMIO region when probing ncsnp!\n");
-		goto out;
-	}
-
-	high = readl_relaxed(base) >> 28;
-	iounmap(base);
-	if (high != 0x1)
-		kvm_ncsnp_support = true;
-
-out:
-	kvm_info("Hisi ncsnp: %s\n", kvm_ncsnp_support ? "enabled" :
-							 "disabled");
-}
diff --git a/arch/arm64/kvm/hisilicon/Kconfig b/arch/arm64/kvm/hisilicon/Kconfig
new file mode 100644
index 000000000000..6536f897a32e
--- /dev/null
+++ b/arch/arm64/kvm/hisilicon/Kconfig
@@ -0,0 +1,7 @@
+# SPDX-License-Identifier: GPL-2.0-only
+config KVM_HISI_VIRT
+	bool "HiSilicon SoC specific virtualization features"
+	depends on ARCH_HISI
+	help
+	  Support for HiSilicon SoC specific virtualization features.
+	  On non-HiSilicon platforms, say N here.
diff --git a/arch/arm64/kvm/hisilicon/Makefile b/arch/arm64/kvm/hisilicon/Makefile
new file mode 100644
index 000000000000..849f99d1526d
--- /dev/null
+++ b/arch/arm64/kvm/hisilicon/Makefile
@@ -0,0 +1,2 @@
+# SPDX-License-Identifier: GPL-2.0-only
+obj-$(CONFIG_KVM_HISI_VIRT) += hisi_virt.o
diff --git a/arch/arm64/kvm/hisilicon/hisi_virt.c b/arch/arm64/kvm/hisilicon/hisi_virt.c
new file mode 100644
index 000000000000..9587f9508a79
--- /dev/null
+++ b/arch/arm64/kvm/hisilicon/hisi_virt.c
@@ -0,0 +1,124 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+/*
+ * Copyright(c) 2022 Huawei Technologies Co., Ltd
+ */
+
+#include <linux/acpi.h>
+#include <linux/of.h>
+#include <linux/init.h>
+#include <linux/kvm_host.h>
+#include "hisi_virt.h"
+
+static enum hisi_cpu_type cpu_type = UNKNOWN_HI_TYPE;
+
+static const char * const hisi_cpu_type_str[] = {
+	"Hisi1612",
+	"Hisi1616",
+	"Hisi1620",
+	"Unknown"
+};
+
+/* ACPI Hisi oem table id str */
+static const char * const oem_str[] = {
+	"HIP06",	/* Hisi 1612 */
+	"HIP07",	/* Hisi 1616 */
+	"HIP08"		/* Hisi 1620 */
+};
+
+/*
+ * Probe Hisi CPU type form ACPI.
+ */
+static enum hisi_cpu_type acpi_get_hisi_cpu_type(void)
+{
+	struct acpi_table_header *table;
+	acpi_status status;
+	int i, str_size = ARRAY_SIZE(oem_str);
+
+	/* Get oem table id from ACPI table header */
+	status = acpi_get_table(ACPI_SIG_DSDT, 0, &table);
+	if (ACPI_FAILURE(status)) {
+		pr_warn("Failed to get ACPI table: %s\n",
+			acpi_format_exception(status));
+		return UNKNOWN_HI_TYPE;
+	}
+
+	for (i = 0; i < str_size; ++i) {
+		if (!strncmp(oem_str[i], table->oem_table_id, 5))
+			return i;
+	}
+
+	return UNKNOWN_HI_TYPE;
+}
+
+/* of Hisi cpu model str */
+static const char * const of_model_str[] = {
+	"Hi1612",
+	"Hi1616"
+};
+
+/*
+ * Probe Hisi CPU type from DT.
+ */
+static enum hisi_cpu_type of_get_hisi_cpu_type(void)
+{
+	const char *model;
+	int ret, i, str_size = ARRAY_SIZE(of_model_str);
+
+	/*
+	 * Note: There may not be a "model" node in FDT, which
+	 * is provided by the vendor. In this case, we are not
+	 * able to get CPU type information through this way.
+	 */
+	ret = of_property_read_string(of_root, "model", &model);
+	if (ret < 0) {
+		pr_warn("Failed to get Hisi cpu model by OF.\n");
+		return UNKNOWN_HI_TYPE;
+	}
+
+	for (i = 0; i < str_size; ++i) {
+		if (strstr(model, of_model_str[i]))
+			return i;
+	}
+
+	return UNKNOWN_HI_TYPE;
+}
+
+void probe_hisi_cpu_type(void)
+{
+	if (!acpi_disabled)
+		cpu_type = acpi_get_hisi_cpu_type();
+	else
+		cpu_type = of_get_hisi_cpu_type();
+
+	kvm_info("detected: Hisi CPU type '%s'\n", hisi_cpu_type_str[cpu_type]);
+}
+
+/*
+ * We have the fantastic HHA ncsnp capability on Kunpeng 920,
+ * with which hypervisor doesn't need to perform a lot of cache
+ * maintenance like before (in case the guest has non-cacheable
+ * Stage-1 mappings).
+ */
+#define NCSNP_MMIO_BASE	0x20107E238
+bool hisi_ncsnp_supported(void)
+{
+	void __iomem *base;
+	unsigned int high;
+	bool supported = false;
+
+	if (cpu_type != HI_1620)
+		return supported;
+
+	base = ioremap(NCSNP_MMIO_BASE, 4);
+	if (!base) {
+		pr_warn("Unable to map MMIO region when probing ncsnp!\n");
+		return supported;
+	}
+
+	high = readl_relaxed(base) >> 28;
+	iounmap(base);
+	if (high != 0x1)
+		supported = true;
+
+	return supported;
+}
diff --git a/arch/arm64/kvm/hisilicon/hisi_virt.h b/arch/arm64/kvm/hisilicon/hisi_virt.h
new file mode 100644
index 000000000000..c4b5acc93fec
--- /dev/null
+++ b/arch/arm64/kvm/hisilicon/hisi_virt.h
@@ -0,0 +1,19 @@
+/* SPDX-License-Identifier: GPL-2.0-or-later */
+/*
+ * Copyright(c) 2022 Huawei Technologies Co., Ltd
+ */
+
+#ifndef __HISI_VIRT_H__
+#define __HISI_VIRT_H__
+
+enum hisi_cpu_type {
+	HI_1612,
+	HI_1616,
+	HI_1620,
+	UNKNOWN_HI_TYPE
+};
+
+void probe_hisi_cpu_type(void);
+bool hisi_ncsnp_supported(void);
+
+#endif /* __HISI_VIRT_H__ */
diff --git a/config.aarch64 b/config.aarch64
index d38f8aa6f6f1..68cbdc285b94 100644
--- a/config.aarch64
+++ b/config.aarch64
@@ -410,6 +410,7 @@ CONFIG_ARM64_NMI=y
 CONFIG_HARDLOCKUP_DETECTOR_PERF=y
 CONFIG_ARM64_HAFT=y
 CONFIG_ARCH_HAS_NONLEAF_PMD_YOUNG=y
+CONFIG_KVM_HISI_VIRT=y
 
 CONFIG_ARM64_4K_PAGES=y
 # CONFIG_ARM64_16K_PAGES is not set
diff --git a/config.aarch64-64k b/config.aarch64-64k
index 0c607f208ba7..235eae609e51 100644
--- a/config.aarch64-64k
+++ b/config.aarch64-64k
@@ -410,6 +410,7 @@ CONFIG_ARM64_NMI=y
 CONFIG_HARDLOCKUP_DETECTOR_PERF=y
 CONFIG_ARM64_HAFT=y
 CONFIG_ARCH_HAS_NONLEAF_PMD_YOUNG=y
+CONFIG_KVM_HISI_VIRT=y
 
 # CONFIG_ARM64_4K_PAGES is not set
 # CONFIG_ARM64_16K_PAGES is not set
-- 
2.43.0

